DOM_task

Tedzury opened the pull request:
### DOM_task

[Demo](https://tedzury.github.io/DOM_kottans_task/) |
[Code base](https://github.com/Tedzury/DOM_kottans_task)

The code is submitted in a dedicated feature branch.

Only code files are submitted.

Please, review.

Tedzury commented:
Hello. Fixed minor issues during self-check. Feel ready to ask for review :)

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,164 @@
+makeFirstPage();
+
+function makeMenu(content) {
+  for (let i = 0; i < content.length; i++) {
```
Use `forEach` instead of a regular cycle when possible.

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,164 @@
+burger.addEventListener("click", menuHandler);
+const menuClasses = menuList.classList;
+
+function menuHandler(e) {
```
You can extract `target` right away with `function menuHandler({ target }) { ...`

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,164 @@
+
+const menuItems = document.querySelectorAll(".menu__item");
+
+for (let i = 0; i < content.length; i++) {
```
You should use as few event listeners in your work as possible, so please try to use only one for the whole menu.

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,164 @@
+}
+
+function makeContent(e) {
+  menuHandler();
```
I am confused, are you redoing your function for the menu every time you go to another page? Isn't `menuHandler` supposed to receive the `e` parameter (tip: it does but it shouldn't)? Why not run this once somewhere?

artem-trubin reviewed (changes requested):
Hi, nice work! Let's make it better.

There is a little problem with functions and their responsibilities. You see, for code to be much more understandable and clearer we should write it as close to its actual meaning as possible. What that means is that if your function is named` makeContent` it is not very clear what exactly is happening there. If we look at your implementation of this function we can see that it does several things:
- it receives the `event` object
- it calls `menuHandler` to reinitiate the menu (which is strange since the menu is static)
- creates a page layout based on the id that is in `event target`
- renders this layout to the page

It's pretty hard to call these actions `Make Content`. 

If we split functions like this into smaller parts that just do their job it becomes much simpler and more clear what is happening in the code. It would be nice to have a function called something like `prepareLayoutById` that gets an ID of a clicked button and **returns** layout. It doesn't need the whole `event` object and it should render to the page right away, it is not its responsibility. You can abstract rendering a prepared layout to a separate function and name it accordingly.

Tedzury commented:
Hello. Hope I fixed everything needed. Removed for loop, removed unneeded listeners for menu items, renamed function makeContent to prepareLayoutById, split this func to two parts, one to get the Id, second to form actual layout.

Few word about func menuHandler: it is responsible for opening and hiding the menu on mobile devices by burger. If user taps the burger - it opens the menu, same time it attaches the event listener to everything that is not a menu, to make it possible for user to close menu if he/she doesn't want to choose new page, just wanna close menu. In case user taps new page - it closes the menu i renders new content. Kind of switcher to menu, I guess.

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,175 @@
+function prepareLayoutById(e) {
+  menuHandler();
+
+  const contentId = getContentId();
```
The whole thing with `getContentId` looks strange. You can just do

```js
const contentId = content.findIndex(item => item.name === target.textContent)
```

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,175 @@
+    return result;
+  }
+
+  function prepareLayout(contentId) {
```
Why does `prepareLayout` has to be a function? You have many of these in code, a function that is just used once, is there a real need for such code to be a function? If you want to group something, maybe just make a comment header.

artem-trubin commented on submissions/Oleksii_Drohachov/DOM_task/script.js:
```diff
@@ -0,0 +1,175 @@
+  }
+
+  function prepareLayout(contentId) {
+    if (contentId != undefined) {
```
Look, you have an indentation for the whole block of code just because of this one `if`. You can redo it as a single-line `if` like this:

```js
if (!contentId) return
```

And just continue your code on the same level of indentation.

artem-trubin reviewed (changes requested):
Hey, let's make some more improvements!

Tedzury commented:
Hello! Fixed everything you mentioned, deleted <ul> in html and created it by JS, to only reach the DOM once. 

In last comment  can't use " if (!contentId) return ", because I have an array, and content with index 0 won't be rendered, instead used " if (contentId === undefined) return  "

Made slight refactoring, hope it makes code better for review. 

Also not sure if got you right in last but one comment: " Why does prepareLayout has to be a function? You have many of these in code, a function that is just used once, is there a real need for such code to be a function? If you want to group something, maybe just make a comment header. "

artem-trubin reviewed (changes requested):
Hey, that's better! But now you've introduced a new bug: your app breaks if you click inside the menu block, but not on the button. I believe you should add a check in your listener function to check if user actually clicked a button.

Tedzury commented:
Good evening. Fixed bug, just replaced "if (contentId === undefined) return"  with  "(contentId === -1 ) return;", it happens because of findIndex, it returns -1 if nothing was found, not undefined.

artem-trubin reviewed (changes requested):
Good find and a fix for a bug! I still have a question about the `prepareLayout` function though, you use it only once right where you define it, why make it a function at all? Functions are supposed to be some group of code lines and actions to either reuse or use later, and here you can just remove the `function` wrapping and nothing would change.

Tedzury commented:
Hello. Sorry for my dumbness. It was really unneeded function. Made it just for logical separation. So, now removed it.

artem-trubin reviewed (approved):
Nice job!
